doc/Precision_Guidelines_for_Ngen_Code_Validation.md

```diff
@@ -0,0 +1,36 @@
+﻿#### Precision guidelines for validation of NGEN code translation and upgrades
+
+**Overview**
+
+Differences in model performance and outcomes often arise when codes are run in different contexts -- i.e., on different operating systems or hardware, with different compilers or compiler settings, under different drivers or wrappers, and when code components are restructured.  The latter context is relevant for the NGEN effort, which is modularizing components of different models into an overarching, flexible framework. There is a need for a standard to determine when simulation differences indicate a coding or implementation error, versus a difference that can be attributed to a method choice or unavoidable cause (e.g., switching from 32-bit to 64-bit hardware, switching solvers).  This document proposes precision standards to be applied when accepting a code reformulation or change.  These are also useful as a real-time simulation check on energy and moisture balances.  An example of such standards being applied is a recent validation of an implementation of a standalone SUMMA model within the NASA Land Information System (LIS) run-time environment.
+  
+**Proposed precision tolerances**
+
+Most components being included in the CFE adopt a double precision (8-byte) standard for internal state and flux variables.  For efficiency, input and output variables may be single (float, 4-byte) precision.  It’s possible that some entire models or components that are included in the future may also have a float precision, which is not uncommon in legacy model code (and is faster than double precision code).  Advanced models may even be able to switch between float and double precision implementations.  Double and float variables are typically now stored with 15- and 7-digit  precision (respectively), but certain operations or uses of the variables may result in an accumulation of error after multiple timesteps, and a loss of matching ability (even when code is correctly implemented) at slightly lower precision levels. _For this reason, 12 digits  is proposed as a standard for validating double values, and 5 digits is proposed as a standard for float variables_ (In practice, the precision for double values could be lower if float variables are involved at any stage of the calculations). While a match at these levels does not entirely rule out rule coding error in a new model version, it would give confidence that the chances of an error are low, and that if an error exists, its impact on simulation outcomes will likely be minimal even in long simulations.  

```

Comment:
>does not entirely rule out rule coding error

does not entirely rule out coding error?

Reply:
It should rules out major ones, perhaps not 100 percent about insignificant ones. Maybe coding error is too strong a word. It guarantees that if any difference exists, its impact on practical applications can be neglected without significant consequences (when certain precision is reached).

Reply:
Sorry, was just trying to point out the typo "rule out rule coding"